pennylane/interfaces/torch.py

```diff
@@ -16,13 +16,69 @@
 compatible quantum nodes.
 """
 # pylint: disable=redefined-outer-name,arguments-differ
+from collections import Iterable
 import inspect
 from functools import partial
+import numbers
 
 import numpy as np
 import torch
+from torch.autograd.function import once_differentiable
 
-from pennylane.utils import unflatten
+
+def _unflatten_torch(flat, model):
+    """Restores an arbitrary nested structure to a flattened Torch tensor.
+
+    See also :func:`_flatten`.
+
+    Args:
+        flat (torch.Tensor): 1D array of items
+        model (array, Iterable, Number): model nested structure
+
+    Raises:
+        TypeError: if ``model`` contains an object of unsupported type
+
+    Returns:
+        Union[torch.Tensor, list, Any], array: first elements of flat arranged into the nested
+        structure of model, unused elements of flat
+    """
+    if isinstance(model, (numbers.Number, str)):
+        return flat[0], flat[1:]
+
+    if isinstance(model, torch.Tensor):
+        idx = model.size
+        res = flat[:idx].view(model.shape)
+        return res, flat[idx:]
+
+    if isinstance(model, Iterable):
+        res = []
+        for x in model:
+            val, flat = _unflatten_torch(flat, x)
+            res.append(val)
+        return res, flat
+
+    raise TypeError("Unsupported type in the model: {}".format(type(model)))
+
+
+def unflatten_torch(flat, model):
+    """Wrapper for :func:`_unflatten_torch`.
+
+    Args:
+        flat (torch.Tensor): 1D array of items
+        model (array, Iterable, Number): model nested structure
+
+    Returns:
+        Union[torch.Tensor, list, Any], array: first elements of flat arranged into the nested
+        structure of model, unused elements of flat
+
+    Raises:
+        ValueError: if ``flat`` has more elements than ``model``
+    """
+    # pylint:disable=len-as-condition
+    res, tail = _unflatten_torch(flat, model)
+    if len(tail) != 0:
+        raise ValueError("Flattened iterable has more elements than the model.")
+    return [torch.as_tensor(i) for i in res]
```

Comment:
I'm not sure what happens at this step?

Reply:
The returned `res` is a list containing either scalars or lists.. This simply casts each list back to a tensor.